lightclient/src/platform/wasm_helpers.rs

```diff
@@ -40,87 +30,381 @@ pub fn sleep(duration: Duration) -> Delay {
+                read_write: read_write::ReadWrite {
+                    now,
+                    incoming_buffer: mem::take(this.read_buffer),
+                    expected_incoming_bytes: if !*this.read_closed { Some(0) } else { None },
+                    read_bytes: 0,
+                    write_bytes_queued,
+                    write_buffers: mem::take(this.write_buffers),
+                    write_bytes_queueable: if !*this.write_closed {
+                        // Limit outgoing buffer size to 128kiB.
+                        // TODO: make configurable?
+                        Some((128 * 1024usize).saturating_sub(write_bytes_queued))
+                    } else {
+                        None
+                    },
+                    wake_up_after: this.read_write_wake_up_after.take(),
+                },
+                read_buffer: this.read_buffer,
+                read_buffer_valid: this.read_buffer_valid,
+                read_buffer_reasonable_capacity: *this.read_buffer_reasonable_capacity,
+                write_buffers: this.write_buffers,
+                write_closed: this.write_closed,
+                close_pending: this.close_pending,
+                read_write_wake_up_after: this.read_write_wake_up_after,
+            })
+        }
+    }
+
+    impl<T> WithBuffers<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        /// Waits until [`WithBuffers::read_write_access`] should be called again.
+        ///
+        /// Returns if an error happens on the socket. If an error happened in the past on the socket,
+        /// the future never yields.
+        pub async fn wait_read_write_again<F>(
+            self: Pin<&mut Self>,
+            timer_builder: impl FnOnce(Instant) -> F,
+        ) where
+            F: future::Future<Output = ()>,
+        {
+            let mut this = self.project();
+
+            // Return immediately if `wake_up_after <= now`.
+            match (&*this.read_write_wake_up_after, &*this.read_write_now) {
+                (Some(when_wake_up), Some(now)) if *when_wake_up <= *now => {
+                    return;
+                }
+                _ => {}
+            }
+
+            let mut timer = pin::pin!({
+                let fut = this
+                    .read_write_wake_up_after
+                    .as_ref()
+                    .map(|when| timer_builder(*when));
+                async {
+                    if let Some(fut) = fut {
+                        fut.await;
+                    } else {
+                        future::pending::<()>().await;
+                    }
+                }
+            });
+
+            // Grow the read buffer in order to make space for potentially more data.
+            this.read_buffer.resize(this.read_buffer.capacity(), 0);
+
+            future::poll_fn(move |cx| {
+                if this.error.is_some() {
+                    // Never return.
+                    return Poll::Pending;
```

Comment:
it's never woken up again this?

Reply:
I believe the smoldot will drop the future entirely if it returns poll::pending, this is similar to smoldot with-buffers